core/iwasm/fast-jit/fe/jit_emit_numberic.c

```diff
@@ -460,16 +465,142 @@ jit_compile_op_i64_arithmetic(JitCompContext *cc, IntArithmetic arith_op,
     return compile_op_int_arithmetic(cc, arith_op, false, p_frame_ip);
 }
 
+static JitReg
+compile_int_and(JitCompContext *cc, JitReg left, JitReg right, bool is_i32)
+{
+    JitReg res;
+
+    /* shortcuts */
+    if (IS_CONST_ZERO(left) || IS_CONST_ZERO(right)) {
+        res = is_i32 ? NEW_CONST(I32, 0) : NEW_CONST(I64, 0);
+        goto shortcut;
+    }
+
+    if (IS_CONST_ALL_ONE(left, is_i32)) {
+        res = right;
+        goto shortcut;
+    }
+
+    if (IS_CONST_ALL_ONE(right, is_i32)) {
+        res = left;
+        goto shortcut;
+    }
+
+    /* do and */
+    res = is_i32 ? jit_cc_new_reg_I32(cc) : jit_cc_new_reg_I64(cc);
+    GEN_INSN(AND, res, left, right);
+
+shortcut:
+    return res;
+}
+
+static JitReg
+compile_int_or(JitCompContext *cc, JitReg left, JitReg right, bool is_i32)
+{
+    JitReg res;
+
+    /* shortcuts */
+    if (IS_CONST_ZERO(left)) {
+        res = right;
+        goto shortcut;
+    }
+
+    if (IS_CONST_ZERO(right)) {
+        res = left;
+        goto shortcut;
+    }
+
+    if (IS_CONST_ALL_ONE(left, is_i32) || IS_CONST_ALL_ONE(right, is_i32)) {
+        res = is_i32 ? NEW_CONST(I32, -1) : NEW_CONST(I64, -1);
+        goto shortcut;
+    }
+
+    /* do or */
+    res = is_i32 ? jit_cc_new_reg_I32(cc) : jit_cc_new_reg_I64(cc);
+    GEN_INSN(OR, res, left, right);
+
+shortcut:
+    return res;
+}
+
+static JitReg
+compile_int_xor(JitCompContext *cc, JitReg left, JitReg right, bool is_i32)
+{
+    JitReg res;
+
+    /* shortcuts */
+    if (left == right) {
+        res = is_i32 ? NEW_CONST(I32, 0) : NEW_CONST(I64, 0);
+        goto shortcut;
+    }
+
+    if (jit_reg_is_const(left) && jit_reg_is_const(right)) {
+        if ((is_i32
+             && (jit_cc_get_const_I32(cc, left)
+                 == jit_cc_get_const_I32(cc, right)))
+            || (!is_i32
+                && (jit_cc_get_const_I64(cc, left)
+                    == jit_cc_get_const_I64(cc, right)))) {
+            res = is_i32 ? NEW_CONST(I32, 0) : NEW_CONST(I64, 0);
```

Comment:
If both left and right are const, just return xor or their const value: left_val ^ right_val

Reply:
seems all bi-operations of integers can use this improvement. shall we apply it to all ?

Reply:
Yes, had better apply it to all